pricegraph/src/num.rs

```diff
@@ -4,9 +4,27 @@ use primitive_types::U256;
 use std::cmp;
 use std::f64;
 
-/// The maximum rounding error, used for asserting that amounts and balances
-/// remain coherent for tests and when built in `debug` profile.
-pub const MAX_ROUNDING_ERROR: f64 = 0.1;
+/// The maximum rounding error for the specified amount, used for asserting that
+/// amounts and balances remain coherent for tests and for `debug` profile.
+///
+/// The maximum rouding error is calcuated by finding the value of the least
+/// significant digit of quantity. This means that quantities can only be off
+/// by that least significant digit.
+pub fn max_rounding_error(quantity: f64) -> f64 {
+    // NOTE: For discussion on the derivation of this formula, see:
+    // https://github.com/gnosis/dex-services/pull/1012#discussion_r440627156
+    const SIGN_EXPONENT_MASK: u64 = 0xfff0_0000_0000_0000;
+    f64::from_bits(quantity.to_bits() & SIGN_EXPONENT_MASK) * f64::EPSILON
+}
+
+/// The maximum rouding error with an epsilon. This is because the assertion
+/// `assert_approx_eq` uses `>` and `<` semantics, while the maximum rounding
+/// error expects `>=` and `<=` semantics.
+#[cfg(test)]
+pub fn max_rounding_error_with_epsilon(quantity: f64) -> f64 {
+    let r = max_rounding_error(quantity);
+    r + max_rounding_error(r)
```

Comment:
Is the intention here to get the next larger float at this point? I found this crate that does the same thing https://docs.rs/float_next_after/0.1.5/float_next_after/trait.NextAfter.html .

Reply:
@e00E Let me know if the documentation is clearer now.